### ISSUE TITLE
Add Apply, Witherable and Align instances

### ISSUE THREAD

**Issue by tysonzero** [michalt/enummapset#14]
I have found the above to be extremely useful for generically working with various types of Maps. Between them you basically get filtering, unioning and intersecting in a convenient and generic way.

I realize this means depending on `witherable`, `these` and `semigroupoids`. But alas the current Haskell orphan instances situation means that there isn't really a good way around that.

**Comment by Mikolaj** [comment]
Hi! I'm your new maintainer. :)

Is it possible to define the instance outside enummapset? Do we expose enough internals? Are the definitions very long? I think I'd rather have another package that depends on enummapset and adds a lot of instances. Some people would not depend on that new package anyway, but just copy-paste only the instances they need, but that's fine. OTOH, they can't copy-paste enummapset easily, so I'd keep its deps small. What do you think?

**Comment by tysonzero** [comment]
It is possible but I mean that has all the usual downsides of orphan instances. I suppose we could also ask the maintainers of the classes to depend on `enummapset` like they depend on `containers` already.

**Comment by Mikolaj** [comment]
A long shot, but surely try talking with maintainers of these packages and perhaps link back to this issue. Talking and being aware of each other is always a good idea. Regarding orphans, it's unfortunate, but currently that's a common thing with many other packages.

**Comment by tysonzero** [comment]
I will create those issues now. Perhaps the long term solution is to convince people that `EnumMap` belongs in containers (as well as or instead of `IntMap`).

We really do need a proper long term language-level solution for orphans though.

**Comment by Mikolaj** [comment]
Agreed on all accounts. Thanks for that. Perhaps separate `IntMap` and `EnumMap` in containers make most sense, because `EnumMap` needs some base operations to inline to, and `IntMap` fits this role perfectly. Also if some people have some particular use case where they use honest, arithmetic `Int`s all over the place, they can underscore this rare situation by using `IntMap` and also they don't need the notation and compilation overhead of `EnumMap Int`.

**Comment by Mikolaj** [comment]
Slightly related: I noticed http://hackage.haskell.org/package/regex-tdfa defines its own `EnumMap` and `EnumSet'. If that was part of containers, they'd surely use the version from base instead of rolling their own.
